### PR TITLE
Feature: Add Kostal RS485 protocol

### PR DESCRIPTION
**Pull request by dalathegreat**
### What
This PR adds support for a new inverter protocol, Kostal RS485

### Why
To add support for Kostal inverters

### CODE REVIEW

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            if (RS485_RXFRAME[i + 1] != KOSTAL_FRAMEHEADER2[i]) {
+              headerB = false;
+            }
+          }
+
+          // "frame B1", maybe reset request, seen after battery power on/partial data
+          if (headerB && (RS485_RXFRAME[6] == 0x5E) && (RS485_RXFRAME[7] == 0xFF)) {
+            send_kostal(frameB1, 10);
+            Serial2.flush();
+            delay(1);
```
Comment (dalathegreat) [inline Software/src/inverter/KOSTAL-RS485.cpp:304]: Rewriting to avoid delay blocks is essential to smooth operation!

**Comment by Maxlado** [comment]
Hi I'm new here and I'm an old mechanical engineer and I'm not very skilled in electronics but I have a Kostal Plenticore plus 8.5 hoked up to an BYD battery, and you were asking for data from some body with this setup. Are there anything I can contribute with or maybe it is to complicated for me to retrieve the things you need?
Regards Anders

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+  } else {
+    float2frame(CyclicData, 0.0, 6);
+  }
+  // Set nominal voltage to value between min and max voltage set by battery (Example 400 and 300 results in 350V)
+  nominal_voltage_dV =
+      (((datalayer.battery.info.max_design_voltage_dV - datalayer.battery.info.min_design_voltage_dV) / 2) +
+       datalayer.battery.info.min_design_voltage_dV);
+  float2frame(BATTERY_INFO, (float)nominal_voltage_dV / 10, 8);
+
+  float2frame(CyclicData, (float)datalayer.battery.info.max_design_voltage_dV / 10, 12);
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:204]: should be offset "10" instead of "12"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+
+  float2frame(CyclicData, (float)average_temperature_dC / 10, 16);
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:206]: should be offset "14" instead of "16"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+  //  Some current values causes communication error, must be resolved, why.
+  float2frame(CyclicData, (float)datalayer.battery.status.current_dA / 10, 20);  // Peak discharge? current (2 byte float)
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:209]: should be offset "18" instead of "20"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+  float2frame(CyclicData, (float)datalayer.battery.status.current_dA / 10, 24);
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:210]: should be offset "22" instead of "24"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+  float2frame(CyclicData, (float)discharge_current_dA / 10, 28);  // BAttery capacity Ah
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:212]: the comment is wrong?  should be "max charge current" at offset "26" (so bytes from 26-29)

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+
+  float2frame(CyclicData, (float)discharge_current_dA / 10, 32);
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:214]: should be offset "34" instead of "32"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+
+  // When SOC = 100%, drop down allowed charge current down.
+
+  if ((datalayer.battery.status.reported_soc / 100) < 100) {
+    float2frame(CyclicData, (float)charge_current_dA / 10, 36);
+  } else {
+    float2frame(CyclicData, 0.0, 36);
+  }
+
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:223]: battery Ah at offset "30" is missing (float32)

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+                      0x4D,   // CRC
+                      0x00};  //
+
+// values in CyclicData will be overwritten at update_modbus_registers_inverter()
+
+uint8_t CyclicData[64] = {
+    0x00,                          // First zero byte pointer
+    0xE2, 0xFF, 0x02, 0xFF, 0x29,  // frame Header
+    0x1D, 0x5A, 0x85, 0x43,  // Current Voltage  (float)                        Modbus register 216, Bytes 6-9
+    0x00, 0x00, 0x8D, 0x43,  // Max Voltage      (2 byte float),                                     Bytes 12-13
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:53]: it's a 4 bytes or 32 bits float from "Bytes 10-13"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+    0x00, 0x00, 0xAC, 0x41,  // BAttery Temperature        (2 byte float)       Modbus register 214, Bytes 16-17
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:54]: it's a 4 bytes or 32 bits float from "Bytes 14-17".  FWIW I checked the ModBus documentation by Kostal on this, where N=2, which means the quantity of registers.  One register is uint16_t, so two are 32bits.

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+    0x00, 0x00, 0x00, 0x00,  // Peak Current (1s period?),  Bytes 18-21 - Communication fault seen with some values (>10A?)
+    0x00, 0x00, 0x00, 0x00,  // Avg current  (1s period?),  Bytes 22-25  - Communication fault seen with some values (>10A?)
+    0x00, 0x00, 0x48, 0x42,  // Max discharge current (2 byte float), Bit 26-29,
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:57]: "Bit" -> "Bytes"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+                             // Sunspec: ADisChaMax
+    0x00, 0x00, 0xC8, 0x41,  // Battery gross capacity, Ah (2 byte float) , Bytes 30-33, Modbus 512
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:59]: 4 byte float

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+    0x00, 0x00, 0xA0, 0x41,  // Max charge current (2 byte float) Bit 36-37, ZERO WHEN SOC=100
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:60]: 4 byte float, "Bit 36-37" -> "Bytes 34-37"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+                             // Sunspec: AChaMax
+    0xCD, 0xCC, 0xB4, 0x41,  // MaxCellTemp (4 byte float) Bit 38-41
+    0x00, 0x00, 0xA4, 0x41,  // MinCellTemp (4 byte float) Bit 42-45
+    0xA4, 0x70, 0x55, 0x40,  // MaxCellVolt  (float), Bit 46-49
+    0x7D, 0x3F, 0x55, 0x40,  // MinCellVolt  (float), Bit 50-53
+
+    0xFE, 0x04,              // Cycle count,
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:67]: let's add a type here, should be "uint16_t"

File: Software/src/inverter/KOSTAL-RS485.cpp [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+    0x00, 0x00, 0xA0, 0x41,  // Max charge current (2 byte float) Bit 36-37, ZERO WHEN SOC=100
+                             // Sunspec: AChaMax
+    0xCD, 0xCC, 0xB4, 0x41,  // MaxCellTemp (4 byte float) Bit 38-41
+    0x00, 0x00, 0xA4, 0x41,  // MinCellTemp (4 byte float) Bit 42-45
+    0xA4, 0x70, 0x55, 0x40,  // MaxCellVolt  (float), Bit 46-49
+    0x7D, 0x3F, 0x55, 0x40,  // MinCellVolt  (float), Bit 50-53
+
+    0xFE, 0x04,              // Cycle count, 
+    0x00,                    // Byte 56
+    0x00,                    // When SOC=100 Byte57=0x40, at startup 0x03 (about 7 times), otherwise 0x02
```
Comment (lewurm) [inline Software/src/inverter/KOSTAL-RS485.cpp:69]: My notes here:
```
Valid bits set in this byte can be:
  - (1 << 0) = 1: ???
  - (1 << 3 | 1 << 4) = 8 | 16: ??? but only comes as pair
  - (1 << 7) = ???
 
 No bits set seem to be a good state.  Not sure why there are traces with other bits set (specifically 1 << 1)
```

Comment (lewurm) [reply Software/src/inverter/KOSTAL-RS485.cpp:69]: also 57=0x39 :-)

**Comment by lewurm** [comment]
> Hi I'm new here and I'm an old mechanical engineer and I'm not very skilled in electronics but I have a Kostal Plenticore plus 8.5 hoked up to an BYD battery, and you were asking for data from some body with this setup. Are there anything I can contribute with or maybe it is to complicated for me to retrieve the things you need? Regards Anders

Hello! We would appreciate some logs around startup.  There are some minimal hints how to do that here: https://github.com/dalathegreat/Battery-Emulator/wiki/Data-needed-for-new-inverter-integration

Basically you need to sniff the RS485 communication between the BYD battery and the Kostal Plenticore Plus.

**Comment by lewurm** [comment]
> Hi I'm new here and I'm an old mechanical engineer and I'm not very skilled in electronics but I have a Kostal Plenticore plus 8.5 hoked up to an BYD battery, and you were asking for data from some body with this setup. Are there anything I can contribute with or maybe it is to complicated for me to retrieve the things you need? Regards Anders

@Maxlado pulling this up again, a bit more concrete instructions on how to obtain a trace specifically with the Plenticore: https://github.com/dalathegreat/Battery-Emulator/wiki/Inverter:-Kostal#traces-for-reverse-engineering

If you have a BYD HVM or HVS that would be very interesting.  Feel free to reach on other channels (for example email: lewurm _AT_ gmail _DOT_ com).

**Pull request merged**